products/db-knownprojects/python/utils.py

```diff
@@ -33,3 +39,111 @@ def psql_insert_copy(table, conn, keys, data_iter):
 
         sql = "COPY {} ({}) FROM STDIN WITH CSV".format(table_name, columns)
         cur.copy_expert(sql=sql, file=s_buf)
+
+
+def hash_each_row(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    e.g. df = hash_each_row(df)
+    this function will create a "uid" column with hashed row values
+    ----------
+    df: input dataframe
+    """
+    df["temp_column"] = df.astype(str).values.sum(axis=1)
+
+    def hash_helper(x):
+        return hashlib.md5(x.encode("utf-8")).hexdigest()
+
+    df["uid"] = df["temp_column"].apply(hash_helper)
+    del df["temp_column"]
+    cols = list(df.columns)
+    cols.remove("uid")
+    cols = ["uid"] + cols
+    return df[cols]
+
+
+def format_field_names(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Change field name to lower case
+    and replace all spaces with underscore
+    """
+
+    def format_func(x):
+        return re.sub(r"\W+", "", x.lower().strip().replace("-", "_").replace(" ", "_"))
+
+    df.columns = df.columns.map(format_func)
+    return df
+
+
+def add_version_date(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Adding today's date as the version of the file
+    -----
+    note that this function is not implemented yet
+    because we might want to get the file creation date
+    directly from the file instead of assigning a date
+    """
+    df["version"] = DATE
+    return df
+
+
+def ETL(func) -> callable:
```

Comment:
Is it necessary for this to be a decorator? I generally think decorators should be reserved for cases where you need to control before/after conditions, e.g. closing a database connector or committing changes.

Reply:
good question, but this is copied over from from the soon-to-be-deprecated KPDB data repo. improvements seem out of scope here

Reply:
ah, sorry, thought this was fresh code. You're right, definitely out of scope.

Reply:
my bad I didn't name / indicate which files were copied over

Reply:
Gotcha I was wondering what was the inspiration for all of this python code